jplag.frontend-utils/src/main/java/jplag/Table.java

```diff
@@ -1,140 +1,78 @@
-            14827, 14831, 14843, 14851, 14867, 14869, 14879, 14887, 14891, 14897, 14923, 14929, 14939, 14947, 14951, 14957, 14969, 14983, 15013,
-            15017, 15031, 15053, 15061, 15073, 15077, 15083, 15091, 15101, 15107, 15121, 15131, 15137, 15139, 15149, 15161, 15173, 15187, 15193,
-            15199, 15217, 15227, 15233, 15241, 15259, 15263, 15269, 15271, 15277, 15287, 15289, 15299, 15307, 15313, 15319, 15329, 15331, 15349,
-            15359, 15361, 15373, 15377, 15383, 15391, 15401, 15413, 15427, 15439, 15443, 15451, 15461, 15467, 15473, 15493, 15497, 15511, 15527,
-            15541, 15551, 15559, 15569, 15581, 15583, 15601, 15607, 15619, 15629, 15641, 15643, 15647, 15649, 15661, 15667, 15671, 15679, 15683,
-            15727, 15731, 15733, 15737, 15739, 15749, 15761, 15767, 15773, 15787, 15791, 15797, 15803, 15809, 15817, 15823, 15859, 15877, 15881,
-            15887, 15889, 15901, 15907, 15913, 15919, 15923, 15937, 15959, 15971, 15973, 15991, 16001};
-    
-    private int[][] data; // in data[][0] stores the number of used elements of every sub-array
-    private int size;
+public class Table {
+    private final Map<Integer, ArrayList<Integer>> mappedEntries;
+    private final int primeNumber;
 
+    /**
+     * Creates the {@link HashMap}.
+     * @param size specifies the initial size and the key mapping (see {@link Table}).
+     */
     public Table(int size) {
-        // Program.print("Length: "+size+"\n", null);
-        int primzahl = 0;
-        if (size >= 16001) {
-            this.size = 16001;   // just to make sure... :-)
-        } else {
-            for (int i = 0; i < prim.length; i++)
-                if ((primzahl = prim[i]) > size)
-                    break;
-            this.size = primzahl;
-        }
-        data = new int[this.size][];
-    }
-
-    public final void add(int hash, int datum) {
-        addTo(hash % size, datum);
+        mappedEntries = new HashMap<>(size);
+        primeNumber = nextPrimeNumber(size);
     }
 
-    public final int[] get(int hash) {
-        return data[hash % size];
+    /**
+     * Returns all stored numbers for a key. Note that all keys with identical <code>(key % prime)</code> are mapped to the
+     * same values (see {@link Table}).
+     * @param key is the specific key.
+     * @return the stored numbers or an empty array if it does not contain the modifiedKey.
+     */
+    public final int[] get(int key) {
+        int actualKey = key % primeNumber;
+        if (!mappedEntries.containsKey(actualKey)) {
+            return null;
+        }
+        List<Integer> result = mappedEntries.get(actualKey);
+        Stream<Integer> stream = Stream.concat(Stream.of(result.size()), result.stream()); // Legacy format: First entry is array length.
+        return stream.mapToInt(Integer::intValue).toArray(); // TODO TS: Requires as the GST algorithm still relies on the array format.
```

Comment:
Note that right now we include some quirks in order to support the legacy format.
Specifically we do not only return the mapped values for the key, we append the number of mapped values as the first value to be returned. The reason for this is that the greedy string tiling algorithm relies on this format where the first entry of an array specifies the number of following entries in the array. This is deeply engrained into the algorithm and needs to be adapted before we can change this method here to only return the mapped values themselves.

Reply:
Adressed by commits below.